File: Solutions/200. Number of Islands/bfs.ts
function numIslands(grid: string[][]): number {
  if (!grid || grid.length === 0) {
    return 0;
  }

  const numberOfRows = grid.length;
  const numberOfColumns = grid[0].length;
  const visitedCells = new Set<string>();
  let islandCount = 0;

  const exploreIsland = (rowIndex: number, columnIndex: number): void => {
    const queue: [number, number][] = [];
    visitedCells.add(`${rowIndex},${columnIndex}`);
    queue.push([rowIndex, columnIndex]);

    while (queue.length > 0) {
      const [queuedRowIndex, queuedColumnIndex] = queue.shift()!;
      const directions = [
        [-1, 0], // UP
        [0, 1], // RIGHT
        [1, 0], // DOWN
        [0, -1], // LEFT
      ];

      for (const direction of directions) {
        const [rowOffset, columnOffset] = direction;
        const contemplatedRowIndex = queuedRowIndex + rowOffset;
        const contemplatedColumnIndex = queuedColumnIndex + columnOffset;

        // Continue if contemplated row index out of grid.
        if (contemplatedRowIndex < 0 || contemplatedRowIndex >= numberOfRows) {
          continue;
        }

        // Continue if contemplated column index out of grid.
        if (
          contemplatedColumnIndex < 0 ||
          contemplatedColumnIndex >= numberOfColumns
        ) {
          continue;
        }

        // Continue if cell value is not "1".
        if (grid[contemplatedRowIndex][contemplatedColumnIndex] !== "1") {
          continue;
        }

        // Continue if cell value is not "1".
        if (
          visitedCells.has(`${contemplatedRowIndex},${contemplatedColumnIndex}`)
        ) {
          continue;
        }

        // Add to queue.
        queue.push([contemplatedRowIndex, contemplatedColumnIndex]);

        // Add to visited cache.
        visitedCells.add(`${contemplatedRowIndex},${contemplatedColumnIndex}`);
      }
    }
  };

  for (let rowIndex = 0; rowIndex < numberOfRows; rowIndex++) {
    for (let columnIndex = 0; columnIndex < numberOfColumns; columnIndex++) {
      if (
        grid[rowIndex][columnIndex] === "1" &&
        !visitedCells.has(`${rowIndex},${columnIndex}`)
      ) {
        exploreIsland(rowIndex, columnIndex);
        islandCount += 1;
      }
    }
  }

  return islandCount;
}

File: Solutions/200. Number of Islands/bfs.py
from typing import List

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid or len(grid) == 0:
            return 0

        numberOfRows = len(grid)
        numberOfColumns = len(grid[0])
        visitedCells = set()
        islandCount = 0

        def exploreIsland(rowIndex: int, columnIndex: int) -> None:
            queue = []
            visitedCells.add((rowIndex, columnIndex))
            queue.append((rowIndex, columnIndex))

            while queue:
                queuedRowIndex, queuedColumnIndex = queue.pop(0)
                directions = [
                    [-1, 0],  # UP
                    [0, 1],   # RIGHT
                    [1, 0],   # DOWN
                    [0, -1]   # LEFT
                ]

                for direction in directions:
                    rowOffset, columnOffset = direction
                    contemplatedRowIndex = queuedRowIndex + rowOffset
                    contemplatedColumnIndex = queuedColumnIndex + columnOffset

                    # Continue if contemplated row index is out of grid
                    if contemplatedRowIndex < 0 or contemplatedRowIndex >= numberOfRows:
                        continue

                    # Continue if contemplated column index is out of grid
                    if contemplatedColumnIndex < 0 or contemplatedColumnIndex >= numberOfColumns:
                        continue

                    # Continue if cell value is not "1"
                    if grid[contemplatedRowIndex][contemplatedColumnIndex] != "1":
                        continue

                    # Continue if cell is already visited
                    if (contemplatedRowIndex, contemplatedColumnIndex) in visitedCells:
                        continue

                    # Add to queue
                    queue.append((contemplatedRowIndex, contemplatedColumnIndex))

                    # Mark as visited
                    visitedCells.add((contemplatedRowIndex, contemplatedColumnIndex))

        for rowIndex in range(numberOfRows):
            for columnIndex in range(numberOfColumns):
                if grid[rowIndex][columnIndex] == "1" and (rowIndex, columnIndex) not in visitedCells:
                    exploreIsland(rowIndex, columnIndex)
                    islandCount += 1

        return islandCount
